kotlin-inject-compiler/ksp/src/main/kotlin/me/tatarka/inject/compiler/ksp/InjectProcessor.kt

```diff
@@ -27,63 +28,69 @@ class InjectProcessor(
     private lateinit var provider: KSAstProvider
     private lateinit var injectGenerator: InjectGenerator
     private lateinit var kmpComponentCreateGenerator: KmpComponentCreateGenerator
-    private var deferredClasses: List<KSClassDeclaration> = mutableListOf()
-    private var deferredFunctions: List<KSFunctionDeclaration> = mutableListOf()
+    private var lastResolver: Resolver? = null
+    private var deferredClassNames: List<KSName> = mutableListOf()
+    private var deferredFunctionNames: List<KSName> = mutableListOf()
 
     private val kmpComponentCreateFunctionsByComponentType = mutableMapOf<AstClass, MutableList<AstFunction>>()
 
     override fun process(resolver: Resolver): List<KSAnnotated> {
+        lastResolver = resolver
         provider = KSAstProvider(resolver, logger)
         injectGenerator = InjectGenerator(provider, options)
         kmpComponentCreateGenerator = KmpComponentCreateGenerator(provider, options)
 
-        val previousDeferredClasses = deferredClasses
-        val previousDeferredFunctions = deferredFunctions
-
-        val componentSymbols = previousDeferredClasses + resolver.getSymbolsWithClassAnnotation(
-            packageName = COMPONENT.packageName,
-            simpleName = COMPONENT.simpleName
-        )
-        deferredClasses = componentSymbols.filterNot { element ->
+        val componentSymbols =
+            resolver.getSymbolsWithAnnotation(COMPONENT.canonicalName).mapNotNull { it as? KSClassDeclaration }
+        val deferredClasses = componentSymbols.filterNot { element ->
             processInject(element, provider, codeGenerator, injectGenerator)
-        }
+        }.toList()
+        deferredClassNames = deferredClasses.mapNotNull { it.qualifiedName }
```

Comment:
If we know that the qualified name must not be null, then I wouldn't use `mapNotNull` and crash for a null value instead. Silently dropping a deferred class sounds scary to me.

Reply:
I originally had that but then decided to drop was ok because this only runs in the case where the symbol gets deferred past the last round (when a dependent symbol is _never_ resolved). This implies other errors in the project so I think it's acceptable to be dropped here if that ever happens.

Reply:
> This implies other errors in the project

Please highlight this by throwing an exception. Silently dropping makes it much harder to diagnose issues.

Reply:
Fair enough, I did tell you the same thing 😅

Reply:
done

Reply:
I still see `mapNotNull`. Any chance this could be changed, too?

Reply:
ooh, if qualifiedName is null? Yeah, sure

Reply:
done, chose to make it a warning since I don't really understand the circumstanced when it might be null and don't want to prevent cases where it is resolvable in a future round.